[써니] 2단계 자동차 경주 제출합니다.

inseonyun opened the pull request:
안녕하세요 리뷰어님! 😊
주말은 잘 보내셨나요!?
지난 번 리뷰, 코멘트 너무너무 감사합니다! :)
금번 PR에서는 지난 번 남겨주신 코멘트와 더불어 좀 더 개선해볼만한 것을 찾아 리펙터링 해봤습니다 ㅎㅎ

1. 코멘트 내용 반영
    1. 입력에 관한 책임은 InputView에 있으니, Controller가 아닌 InputView에서 처리
    > 기존 컨트롤러에서 입력 값을 받아와 try-catch로 예외를 처리해 해당 함수를 재귀 호출하는 방식에서, InputView에서 validate 검증 결과에 따라 반복 입력 받도록 수정하였습니다.
    2. 클래스 내에서만 사용되는 동반 객체 접근 제한 수정!
    > 기존 접근 제한이 없던 상수들을 private으로 접근 제한하였습니다!
    3. 테스트명 수정
    > 테스트 함수가 무엇을 수행하고 테스트하는지 알 수 있도록 함수명을 수정하였고, 테스트 클래스 명에서 알 수 있는 '테스트'라는 단어는 함수명에 넣지 않도록 수정하였습니다!
    4. 조건문을 통한 throw가 아닌 require를 이용한 요구 조건 validate 수행

코멘트 기다리고 있겠습니다! 감사합니다! 😊

inseonyun committed:
refactor: companion object 상수 접근 제한자 수정

inseonyun committed:
refactor: RacingCarGameTest 테스트 함수명 수정

inseonyun committed:
refactor: TryCountValidation check 함수 수정

inseonyun committed:
refactor: NullValidation check 함수 수정

inseonyun committed:
refactor: NameValidation check 함수 수정

inseonyun committed:
refactor: OutputView에서 라운드 결과 및 최종 우승자 출력하도록 수정

inseonyun committed:
refactor: 입력 예외 발생 재귀 호출 수정

inseonyun committed:
refactor: NameValidation Regex Pattern 상수화

inseonyun committed:
refactor: NullValidation 클래스, 함수명 변경

inseonyun committed:
refactor: InputView에서 입력 값 예외 발생 시 반복 입력 처리

1. NameValidation, TryCountValidation - validate 함수에서 검증 결과 boolean 반환
2. InputView에서 각 검증 결과에 따라 반복 입력 받도록 수정
3. Controller input- 함수명 init-으로 수정

inseonyun committed:
refactor: 테스트 함수명 수정

inseonyun committed:
refactor: 가장 멀리간 자동차 반환 함수 수정

1. getMaxPositionCarNames - List<Car>에서 name으로 매핑해 List<String> 반환
2. printWinners - subString이 아닌 joinToString을 이용해 출력 형식 맞춤

BeokBeok commented on src/main/kotlin/service/RacingCarGameService.kt:
```diff
@@ -23,31 +23,15 @@ class RacingCarGameService {
-
-        carsInfo.forEach { car ->
-            winners += "${car.name}, "
+        return carsInfo.filter { maxCarInfo.isSamePosition(it) }.map {
+            car ->
+            car.name
         }
```
```suggestion
        return carsInfo.filter { maxCarInfo.isSamePosition(it) }
            .map { car ->
                car.name
            }
```
람다같은 경우에는 "변수명 ->" 까지는 같은 라인에 작성하는 것이 좋습니다
ref. https://kotlinlang.org/docs/coding-conventions.html#lambdas

BeokBeok commented on src/main/kotlin/validation/NameValidation.kt:
```diff
@@ -1,17 +1,28 @@
 package validation
 
+import view.OutputView
```
```suggestion
```
Validation은 Controller가 아닙니다.
즉, View와 관련된 의존성을 가지면 안됩니다.
OutView의 의존성을 가지지 않도록 코드를 개선해보면 어떨까요?

BeokBeok commented on src/main/kotlin/validation/NullOrBlankValidation.kt:
```diff
@@ -0,0 +1,13 @@
+        return input
+    }
+
+    private const val INPUT_IS_EMPTY_ERROR_MESSAGE = "[ERROR] 아무것도 입력하지 않았습니다."
```
클래스의 레이아웃은 대체로 변수 -> 함수 형태로 작성하는 것이 컨벤션입니다.

ref. https://kotlinlang.org/docs/coding-conventions.html#class-layout

BeokBeok commented on src/main/kotlin/view/InputView.kt:
```diff
@@ -1,22 +1,29 @@
-        return carNames
+        var input = ""
+        var isValidate = false
+        while (!isValidate) {
+            OutputView.printCar()
+            input = readln()
+            isValidate = NameValidation.validateCarName(input)
+        }
```
이름이 유효할때까지 계속해서 반복하고 있네요.
InputView가 어떤 책임을 갖는지 다시 한번 생각해보시면 좋을 것 같아요.
InputView는 입력값을 받아서 전달하는 역할을 할 뿐이지, 올바른 입력값을 받을때까지 계속해서 호출하는 것과는 별개입니다.
입력값을 받아서 확인하는 곳이 어디인지 확인해보시고 그 곳에서 처리하도록 개선해보면 어떨까요? (이하 관련 내용 동일)

inseonyun replied:
아무래도 해당 코멘트 부분에서 도무지 생각이 나질 않네요....
입력값을 확인하는 NameValidation에서 이를 처리하고 싶은데, 반복 입출력을 위해선 InputView와 OutputView에 접근을 하게 되어 View에 대한 의존성이 생깁니다...이러한 문제를 해결하기 위해 Validation 의 생성자에 View 클래스를 주입 하는 것이 맞는 판단일까요..? 😂

BeokBeok replied:
힌트를 조금 더 드리겠습니다.
inputName()을 호출한 곳에서 반환된 값을 판단합니다.
만약 값이 유효하면 다음 프로세스를 진행하고, null과 같은 유효하지 않은 값이라면 해당 함수를 다시 한번 호출해주면 됩니다.
참고로 프로그래밍 요구사항에는 "유효한 값을 받을때까지 프로그램이 종료되지 않는다"와 같은 항목은 없습니다. 😅

BeokBeok commented on src/test/kotlin/RacingCarGameTest.kt:
```diff
@@ -7,25 +7,25 @@ class RacingCarGameTest {
 
     @Test
-    fun `우승한 자동차를 정상적으로 출력하는지 테스트`() {
+    fun `가장 멀리간 자동차들을 확인한다`() {
```
테스트 함수명이 위에 작성된 함수명과 너무 유사해서 햇갈릴 여지가 있네요.
"공동 우승" 과 관련된 문구를 추가해보면 어떨까요?

BeokBeok reviewed (changes requested):
이전 단계 피드백과 자동차 경주 리팩터링 미션하시느라 고생하셨습니다. 👍 
고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪

inseonyun committed:
refactor: getMaxPositionCarNames() 람다식 수정

inseonyun committed:
refactor: 테스트 함수명 수정

inseonyun committed:
refactor: 변수 선언 위치 수정

inseonyun committed:
refactor: 입력값 검증 Controller에서 수행하도록 수정

1. InputView는 오직 입력만 받는다
2. Validation는 입력값에 대한 검증만 이루어진다
3. Controller에서 위 값을 검증하여 예외 혹은 게임 진행을 한다

BeokBeok commented on src/main/kotlin/validation/TryCountValidation.kt:
```diff
@@ -1,26 +1,29 @@
+    private const val MINIMUM_TRY_COUNT = 1
+
+    private const val TRY_COUNT_INTEGER_ERROR_MESSAGE = "[ERROR] 시도 횟수는 숫자만 입력해 주세요"
+    private const val TRY_COUNT_POSITIVE_ERROR_MESSAGE = "[ERROR] 시도 횟수는 양의 정수여야 합니다"
+    fun validate(input: String) {
```
```suggestion
    private const val TRY_COUNT_POSITIVE_ERROR_MESSAGE = "[ERROR] 시도 횟수는 양의 정수여야 합니다"
    
    fun validate(input: String) {
```
변수 정의부의 마지막과 함수 정의부 초반부에는 개행으로 구분해주시는 것이 좋습니다. (이하 관련 내용 동일)

BeokBeok reviewed (approved):
피드백 반영하시느라 고생하셨습니다. 👍 
인선님과 미션하면서 저 또한 성장하게 된 것 같습니다. 😄 
좋은 개발자가 되기 위한 소중한 경험이 되었길 바랍니다. 💪
